050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,832 @@
+Since Apache Kafka plans to remove ZooKeeper support in its 4.0 release, supporting KRaft mode in all its different forms is a requirement for us.
+
+### Stretch clusters
+
+There seems to be increased demand Kafka clusters stretched across multiple Kubernetes clusters.
+Apache Kafka itself has latency limitations and might not run well when stretched over distant clusters with big latency.
+But there are many use cases where stretched clusters might be useful.
+For example:
+
+* When running closely colocated Kubernetes clusters
+* When migrating from old Kubernetes cluster to new Kubernetes cluster
+
+## Proposal
+
+This proposal suggests introduction of Kafka _node pools_.
+_Node pools_ are groups of Kafka nodes which will have the same configuration.
+A Kafka cluster can consist of one or more pools.
+Each node pool can belong to only one Kafka cluster.
+
+### `KafkaNodePool` custom resource
+
+A new custom resource called `KafkaNodePool` will be added to Strimzi.
+One `KafkaNodePool` resource will represent one _node pool_.
+It will take over some of the configuration from the `Kafka` CR, reuse other parts from it and add some new configuration options as well.
+The `KafkaNodePool` resource will be a _public_ resource.
+It will be created and managed directly by users (unlike for example the `StrimziPodSet` resource which is used only internally by Strimzi and should not be touched by the users).
+
+The _cluster-wide_ configuration options will remain part of the `Kafka` CR.
+While the _pool-specific_ options will be taken from `KafkaNodePool` resources
+
+For the initial implementation, the following options are proposed for the `KafkaNodePool` resource and its `.spec` section:
+
+* `replicas` indicating the number of nodes in the pool
+* `storage` indicating the storage of the nodes in this pool
+* `resources` indicating the amount of resources (requests / limits) assigned to the nodes in this pool
+* `template` for customizing the pods and other resources which are part of the nodes in this pool
+* `jvmOptions` for customizing the JVM configuration
+
+For example:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: my-pool
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  resources:
+    requests:
+      memory: 8Gi
+      cpu: 2
+    limits:
+      memory: 8Gi
+      cpu: 4
+  jvmOptions:
+    -Xms: 2048m
+    -Xmx: 2048m
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: true
+  template:
+    podSet:
+      metadata:
+        labels:
+          key1: label1
+    pod:
+      terminationGracePeriodSeconds: 60
+```
+
+The replicas and storage sections will be required in `KafkaNodePool`.
+The other fields will be optional.
+
+When the optional field is not present in `KafkaNodePool` and is present in `Kafka.spec.kafka`, it will default to the configured value from the `Kafka` CR.
```

Comment:
I do not find this useful. It just adds more complexity while I think nobody will really follow this. Also, what if the options is configured by the default default? (i.e. the default value for the default value in Kafka CR?)

Reply:
> Also, what if the options is configured by the default default? (i.e. the default value for the default value in Kafka CR?)

I think you are referring to things like LimiRanges. That's different because they are only applied at the pod admission stage and changing values would only have an effect on new pods. Instead, changing Kafka resources would cause cause a rolling update.

Reply:
No. There are fields in the Kafka CR which do have their own defaults when not set. So I think what you suggest here is not really as straightforward and understandable. In any case, it is definitely not something what is key to the proposal. If we see the need for this or some misunderstandings of what values are used, this can be added at any point later.

Reply:
I agree that this is not key to the proposal and the above status would be confusing, but yet this is a behavior that could be unexpected for some users and make troubleshooting more complicated (you need to combine two different resources). An alternative could be to simply log the KafkaNodePool applied configuration at startup. WDYT?

Reply:
I do not see the confusion TBH. But the advantage of this being not a key part of the proposal is that if there actually turns out to be any confusion by the users, then we can decide what to add and where exactly.